pkg/controller/nodecsrapprover/node_csr_approver.go

```diff
@@ -130,40 +153,121 @@ func (r *reconciler) reconcile(ctx context.Context, request reconcile.Request) e
-func validateCSRRequest(csr *certificatesv1beta1.CertificateSigningRequest) error {
-	csrBlock, rest := pem.Decode(csr.Spec.Request)
-	if csrBlock == nil {
-		return fmt.Errorf("no certificate request found for the given CSR")
+	// Ensure system:nodes and system:authenticated are in groups
+	if len(csr.Spec.Groups) < 2 {
+		return "", fmt.Errorf("there are less than 2 groups")
 	}
-	if len(rest) != 0 {
-		return fmt.Errorf("found more than one PEM encoded block in the result")
+	if !sets.NewString(csr.Spec.Groups...).HasAll(nodeGroup, authenticatedGroup) {
+		return "", fmt.Errorf("'%s' and/or '%s' are not in its groups", nodeGroup, authenticatedGroup)
 	}
-	csrRequest, err := x509.ParseCertificateRequest(csrBlock.Bytes)
-	if err != nil {
-		return err
+
+	// Check are present usages matching allowed usages
+	if len(csr.Spec.Usages) != 3 {
+		return "", fmt.Errorf("there are no exactly three usages defined")
+	}
+	for _, usage := range csr.Spec.Usages {
+		if !isUsageInUsageList(usage, allowedUsages) {
+			return "", fmt.Errorf("usage %v is not in the list of allowed usages (%v)", usage, allowedUsages)
+		}
 	}
+
+	return nodeName, nil
+}
+
+// validateX509CSR validates the certificate request by comparing CN with username,
+// and organization with groups.
+func (r *reconciler) validateX509CSR(csr *certificatesv1beta1.CertificateSigningRequest, certReq *x509.CertificateRequest, machine v1alpha1.Machine) error {
 	// Validate Subject CommonName
-	if csrRequest.Subject.CommonName != csr.Spec.Username {
-		return fmt.Errorf("commonName '%s' is different then CSR username '%s'", csrRequest.Subject.CommonName, csr.Spec.Username)
+	if certReq.Subject.CommonName != csr.Spec.Username {
+		return fmt.Errorf("commonName '%s' is different then CSR username '%s'", certReq.Subject.CommonName, csr.Spec.Username)
 	}
+
 	// Validate Subject Organization
-	if len(csrRequest.Subject.Organization) != 1 {
-		return fmt.Errorf("error validating subject organizations")
+	if len(certReq.Subject.Organization) != 1 {
+		return fmt.Errorf("expected only one organization but got %d instead", len(certReq.Subject.Organization))
+	}
+	if certReq.Subject.Organization[0] != nodeGroup {
+		return fmt.Errorf("organization '%s' doesn't match node group '%s'", certReq.Subject.Organization[0], nodeGroup)
+	}
+
+	// Validate SAN DNS names
+	for _, dns := range certReq.DNSNames {
+		if len(dns) == 0 {
+			continue
+		}
+
+		var foundSAN bool
+		for _, addr := range machine.Status.Addresses {
+			if addr.Address == dns {
+				foundSAN = true
+				break
+			}
+		}
+		if !foundSAN {
+			// Some providers don't have reachable DNS names and those are not in addresses.
+			// In that case we check is the DNS name matching the node name, and if yes we
+			// assume the DNS name is valid and allowed.
+			if machine.Status.NodeRef.Name == dns {
+				continue
+			}
+			return fmt.Errorf("dns name '%s' cannot be associated with node '%s'", dns, machine.Status.NodeRef.Name)
+		}
 	}
```

Comment:
Not a blocker, but why not creating a set of addresses to avoid the nested iteration and make it more compact? Something like:
```
	machineAddressSet := sets.NewString(machine.Status.NodeRef.Name)
	for _, addr := range machine.Status.Addresses {
		machineAddressSet.Insert(addr.Address)
	}
	// Validate SAN DNS names
	for _, dns := range certReq.DNSNames {
		if len(dns) == 0 {
			continue
		}
		if !machineAddressSet.Has(dns) {
			return fmt.Errorf("dns name '%s' cannot be associated with node '%s'", dns, machine.Status.NodeRef.Name)
		}
	}
```

Reply:
Fixed.